### PR TITLE
Multitenancy for MAM/MUC

### PR DESCRIPTION
**Pull request by arcusfelis**
This PR addresses "MAM/MUC test suites are passing for dynamic domains".

Proposed changes include:
* hook refactoring

* Common Tests Hooks for big tests fail, if init_per_suite fails. Added some more strictness there. Separate PR: https://github.com/esl/MongooseIM/pull/3133
* More granular rules in domain_isolation_SUITE, also converted it for HostTypes.

* We COULD pass hosttype as a separate arg to muclight DB backend (and we probably should), but it would increase complexity of the current PR, I would like to do it in a separate one.

* ejabberd_users refactoring would be a separate PR.

* REST API should be done in a separate PR.

* commands API should be done in a separate PR.

Moved into #3137:
* Add reason to some XMPP and HTTP errors. 
* New module mongoose_record_pp to pretty print records in logs (it should be called manually though).

### CODE REVIEW

File: big_tests/dynamic_domains.config [hunk 1]
```diff
@@ -38,5 +56,121 @@
         {username, <<"bOb">>},
         {server, <<"domain.example.com">>},
         {host, <<"localhost">>},
-        {password, <<"makrolika">>}]}
+        {password, <<"makrolika">>}]},
+    {carol, [
```
Comment (chrzaszcz) [inline big_tests/dynamic_domains.config:60]: Please only add users that are actually used by the enabled test suites.

Comment (arcusfelis) [reply big_tests/dynamic_domains.config:60]: done

File: big_tests/dynamic_domains.config [hunk 2]
```diff
@@ -1,11 +1,21 @@
...
+{disable_disco_tests, true}.
+{disable_s2s_tests, true}.  %% Requires mod_register
+{disable_carbons_tests, true}.
+{disable_roster_tests, true}.
+
 {hosts, [{mim,  [{node, mongooseim@localhost},
                  {domain, <<"domain.example.com">>},
+                 {host_type, <<"test type">>},
                  {secondary_domain, <<"domain.example.org">>},
+                 {secondary_host_type, <<"test type">>},
```
Comment (NelsonVides) [inline big_tests/dynamic_domains.config:13]: Why's this the same than the first host type?

File: big_tests/src/ct_mongoose_log_hook.erl [hunk 3]
```diff
@@ -163,28 +163,28 @@ make_content(CurrentLineNum, Line) ->
...
-            ct:pal("issue=\"ct_mongoose_log_hook created log file\", "
-                   "ct_node=~p, reader=~p, reader_node=~p, writer=~p, out_file=~p",
-                   [node(), Reader, node(Reader), Writer, OutFile]),
-            State#state{reader=Reader, writer=Writer, out_file=OutFile,
-                        current_line_num=CurrentLineNum, url_file=UrlFile};
-        Reason ->
-            ct:pal("issue=\"Failed to init ct_mongoose_log_hook\", node=~p, reason=~p",
-                   [Node, Reason]),
-            State#state{out_file=OutFile}
+    try
```
[the remark this answers is not in the document]
Comment (arcusfelis) [reply big_tests/src/ct_mongoose_log_hook.erl:172]: it fails here when suite init_per_suite fails. Fails because RunDir directory is not created at this point.

File: big_tests/dynamic_domains.config (hunk 2 above)
Comment (arcusfelis) [reply big_tests/dynamic_domains.config:13]: idk, it works like that xdd
I think it's a value, you get if you try to convert "domain.example.org" to host type.

File: big_tests/tests/muc_SUITE.erl [hunk 4]
```diff
@@ -85,15 +84,37 @@
...
+all() -> 
+    case disco_disabled() of
+        true ->
+            [];
+        false ->
+            [{group, disco},
+             {group, disco_non_parallel},
+             {group, disco_rsm},
+             {group, disco_rsm_with_offline}]
+    end ++ [
```
Comment (NelsonVides) [inline big_tests/tests/muc_SUITE.erl:117]: What about doing a trick like Denys did at https://github.com/esl/MongooseIM/blob/5166ab32fbeef8f1e369c2ec3b3e74a8d2114663/big_tests/dynamic_domains.spec#L15 seems clearer, smaller diff, doesn't make a mess of the test-suite, and it's easier to add and remove in the future 🤔

File: big_tests/tests/muc_SUITE.erl [hunk 5]
```diff
@@ -319,16 +341,20 @@ init_per_suite(Config) ->
     %% For mocking with unnamed functions
     mongoose_helper:inject_module(?MODULE),
     Config2 = escalus:init_per_suite(Config),
-    Config3 = dynamic_modules:save_modules(domain(), Config2),
-    load_muc(muc_host()),
+    Config3 = dynamic_modules:save_modules(host_type(), Config2),
+    HostType = ct:get_config({hosts, mim, host_type}),
```
Comment (NelsonVides) [inline big_tests/tests/muc_SUITE.erl:345]: You've just defined a helper `host_type()` function up there, could be defined in `domains_helper` and be used everywhere around big_tests btw.

File: src/muc_light/mod_muc_light_db_rdbms.erl [hunk 6]
```diff
@@ -774,28 +773,27 @@ modify_aff_users_transaction(MainHost, RoomUS, RoomID, AffUsersChanges,
...
 %% ------------------------ Common ------------------------
 
--spec main_host(JIDOrServer :: jid:simple_bare_jid() | binary()) -> jid:lserver().
+-spec main_host(JIDOrServer :: jid:simple_bare_jid() | binary()) -> mongooseim:host_type().
 main_host({_, RoomS}) ->
     main_host(RoomS);
 main_host(MUCServer) ->
-    {ok, MainHost} = mongoose_subhosts:get_host(MUCServer),
-    MainHost.
+    mod_muc_light_utils:muc_host_to_host_type(MUCServer).
```
Comment (NelsonVides) [inline src/muc_light/mod_muc_light_db_rdbms.erl:799]: I think that this function can change its name, it doesn't really reflect anymore what it used to do.

File: big_tests/tests/muc_SUITE.erl (hunk 4 above)
Comment (arcusfelis) [reply big_tests/tests/muc_SUITE.erl]: done

File: src/mongoose_api_client.erl [hunk 7]
```diff
@@ -63,6 +63,8 @@ cowboy_router_paths(Base, _Opts) ->
 
 init(Req, Opts) ->
     Bindings = maps:to_list(cowboy_req:bindings(Req)),
+    ?LOG_ERROR(#{what => rest_init,
+                 cowboy_req => Req}),
```
Comment (NelsonVides) [inline src/mongoose_api_client.erl:67]: `ERROR`? This seems like a debug left-over

File: src/mongoose_api_client.erl [hunk 8]
```diff
@@ -112,6 +114,8 @@ delete_resource(Req, #http_api_state{command_category = Category,
 to_json(Req, #http_api_state{command_category = Category,
                              command_subcategory = SubCategory,
                              bindings = B} = State) ->
+    ?LOG_ERROR(#{what => rest_to_json,
+                 cowboy_req => Req}),
```
Comment (NelsonVides) [inline src/mongoose_api_client.erl:118]: Likewise here

File: src/mongoose_api_client.erl [hunk 9]
```diff
@@ -122,6 +126,8 @@ to_json(Req, #http_api_state{command_category = Category,
 from_json(Req, #http_api_state{command_category = Category,
                                command_subcategory = SubCategory,
                                bindings = B} = State) ->
+    ?LOG_ERROR(#{what => rest_from_json,
+                 cowboy_req => Req}),
```
Comment (NelsonVides) [inline src/mongoose_api_client.erl:130]: And similarly

File: src/mongoose_api_common.erl [hunk 10]
```diff
@@ -327,6 +327,8 @@ error_response(ErrorType, Reason, Req, State) ->
                     L when is_list(L) -> list_to_binary(L);
                     Other -> list_to_binary(io_lib:format("~p", [Other]))
                 end,
+    ?LOG_ERROR(#{what => rest_common_error,
+                 error_type => ErrorType, reason => Reason, cowboy_req => Req}),
```
Comment (NelsonVides) [inline src/mongoose_api_common.erl:331]: Now the only thing that concerns me is the fact that we log here `cowboy_req`. Doesn't that might contain private data? We might leak here.

File: src/mongoose_client_api/mongoose_client_api.erl [hunk 11]
```diff
@@ -77,6 +78,14 @@ set_resp_body_if_missing(Body, Req) ->
             cowboy_req:set_resp_body(Body, Req)
     end.
 
+maybe_report_error(StatusCode, Req, Body) when StatusCode >= 400 ->
+    ?LOG_WARNING(#{what => reply_error,
+                   stacktrace => erlang:process_info(self(), current_stacktrace),
+                   status_code => StatusCode,
+                   cowboy_req => Req, body => Body});
```
Comment (NelsonVides) [inline src/mongoose_client_api/mongoose_client_api.erl:85]: Same issue I had before, logging the `cowboy_req` and the `body` might leak private information.

File: src/mongoose_client_api/mongoose_client_api_rooms.erl [hunk 12]
```diff
@@ -111,20 +111,29 @@ from_json(Req, State) ->
...
-            {false, Req, State}
+            mongoose_client_api:bad_request(Req, <<"Failed to parse parameters">>, State)
     end.
 
 handle_request(Method, JSONData, Req, State) ->
     case handle_request_by_method(Method, JSONData, Req, State) of
-        {error, _, _} ->
-            {false, Req, State};
+        {error, Short, Desc} = Other ->
+            mongoose_client_api:bad_request(Req, format_error(Short, Desc), State);
```
Comment (NelsonVides) [inline src/mongoose_client_api/mongoose_client_api_rooms.erl:120]: Can we have all this "better reporting" in a separate PR? This one is sufficiently enormous, and rest reporting errors is a different topic. There's also the question of whether we want to log the entire requests as I mentioned before, which I doubt we want. So these things could be moved to a different PR where we discuss such thing without interrupting mam's multitenancy.

File: src/muc_light/mod_muc_light_room_config.erl [hunk 13]
```diff
@@ -73,7 +73,7 @@
 %%====================================================================
 
 -spec schema_from_definition(UserDefinedSchema :: user_defined_schema()) -> schema().
-schema_from_definition(UserDefinedSchema) ->
+schema_from_definition(UserDefinedSchema) when is_list(UserDefinedSchema) ->
```
Comment (NelsonVides) [inline src/muc_light/mod_muc_light_room_config.erl:76]: Why this numb change? `lists:foldl` will already check that it is a list, and this function should then be on the stacktrace. Did that go wrong somewhere?

File: src/domain/mongoose_lazy_routing.erl [hunk 14]
```diff
@@ -408,9 +422,12 @@ check_that_domain_and_iq_match({#iq_table_key{host_type = HostType,
                                {_, {HostType, Pattern}}) when is_binary(Namespace),
                                                               Component =/= '_' ->
     true;
-check_that_domain_and_iq_match(IQ, Domain) ->
+check_that_domain_and_iq_match({Key=#iq_table_key{}, _} = IQEntry, Domain) ->
```
Comment (NelsonVides) [inline src/domain/mongoose_lazy_routing.erl:425]:
```suggestion
check_that_domain_and_iq_match({Key = #iq_table_key{}, _} = IQEntry, Domain) ->
```

File: src/ejabberd_c2s.erl [hunk 15]
```diff
@@ -2276,7 +2276,8 @@ process_privacy_iq(Acc1, To, StateData) ->
             From = mongoose_acc:from_jid(Acc2),
             {Acc3, NewStateData} = process_privacy_iq(Acc2, Type, To, StateData),
             Res = mongoose_acc:get(hook, result,
-                                   {error, mongoose_xmpp_errors:feature_not_implemented()}, Acc3),
+                                   {error, mongoose_xmpp_errors:feature_not_implemented(
+                                             <<"en">>, <<"Privacy IQ from c2s">>)}, Acc3),
```
Comment (NelsonVides) [inline src/ejabberd_c2s.erl:2280]: Not sure this error message is very meaningful for a client. Maybe explain something about what was violated here.

File: src/ejabberd_local.erl [hunk 16]
```diff
@@ -114,7 +114,8 @@ process_iq(#iq{ xmlns = XMLNS } = IQ, Acc, From, To, _El) ->
         [{_, IQHandler}] ->
             gen_iq_component:handle(IQHandler, Acc, From, To, IQ);
         [] ->
-            ejabberd_router:route_error_reply(To, From, Acc, mongoose_xmpp_errors:feature_not_implemented())
+            ejabberd_router:route_error_reply(To, From, Acc,
+                mongoose_xmpp_errors:feature_not_implemented(<<"en">>, <<"From local handler">>))
```
Comment (NelsonVides) [inline src/ejabberd_local.erl:118]: This error message, could be something like... `Local server does not implement this feature`, where the keyboard is the word `Local`. Should be a string that means something to the end user, and that points us to the place in code that generated it, at the same time.

Comment (NelsonVides) [reply src/ejabberd_local.erl:118]: But again, maybe all these extended error messages could also be moved to a separate PR, this is a different topic to mam.

File: src/ejabberd_sm.erl [hunk 17]
```diff
@@ -540,15 +541,21 @@ handle_info({route, From, To, Packet}, State) ->
     route(From, To, Packet),
     {noreply, State};
 handle_info({register_iq_handler, Host, XMLNS, IQHandler}, State) ->
-    ets:insert(sm_iqtable, {{XMLNS, Host}, IQHandler}),
+    try
+        ets:insert_new(sm_iqtable, {{XMLNS, Host}, IQHandler})
```
[the remark this answers is not in the document]
Comment (arcusfelis) [reply src/ejabberd_sm.erl:545]: good

File: src/muc_light/mod_muc_light_room_config.erl (hunk 13 above)
Comment (arcusfelis) [reply src/muc_light/mod_muc_light_room_config.erl:76]: It makes stacktraces better, failing in foldl is less pretty.
Function clause on schema_from_definition is better.

(it's here because test cases use that function).

File: src/mongoose_client_api/mongoose_client_api.erl (hunk 11 above)
Comment (arcusfelis) [reply src/mongoose_client_api/mongoose_client_api.erl:85]: Users would have to write log filters to filter out private info from the request.
We have cowboy_req field in docs.

File: src/mongoose_api_client.erl (hunk 7 above)
Comment (arcusfelis) [reply src/mongoose_api_client.erl:67]: leftovers

File: src/mod_smart_markers.erl [hunk 18]
```diff
@@ -144,8 +144,7 @@ get_chat_markers(ChatType, #jid{lserver = LServer} = To, Thread, TS) ->
     Host = case ChatType of
                one2one -> LServer;
                groupchat ->
-                   {ok, H} = mongoose_subhosts:get_host(LServer),
-                   H
+                   mod_muc_light_utils:muc_host_to_host_type(LServer)
```
[the remark this answers is not in the document]
Comment (arcusfelis) [reply src/mod_smart_markers.erl:147]: I don't know how, but we have test cases for chat markers in suites, which need this.

File: src/mam/mod_mam.erl [hunk 19]
```diff
@@ -160,64 +161,57 @@
...
-    %% `parallel' is the only one recommended here.
-    IQDisc = gen_mod:get_opt(iqdisc, Opts, parallel), %% Type
-    [mod_disco:register_feature(Host, Feature) || Feature <- features(?MODULE, Host)],
-    gen_iq_handler:add_iq_handler(ejabberd_sm, Host, ?NS_MAM_04,
-                                  ?MODULE, process_mam_iq, IQDisc),
-    gen_iq_handler:add_iq_handler(ejabberd_sm, Host, ?NS_MAM_06,
-                                  ?MODULE, process_mam_iq, IQDisc),
-    ejabberd_hooks:add(hooks(Host)),
-    ensure_metrics(Host),
+    ?LOG_ERROR(#{what => mam_starting, host_type => HostType}),
```
Comment (NelsonVides) [inline src/mam/mod_mam.erl:201]: Leftovers here

File: src/mam/mod_mam.erl [hunk 20]
```diff
@@ -160,64 +161,57 @@
...
--spec stop(Host :: jid:server()) -> any().
-stop(Host) ->
+-spec stop(HostType :: host_type()) -> any().
+stop(HostType) ->
     ?LOG_INFO(#{what => mam_stopping}),
-    ejabberd_hooks:delete(hooks(Host)),
-    gen_iq_handler:remove_iq_handler(ejabberd_sm, Host, ?NS_MAM_04),
-    gen_iq_handler:remove_iq_handler(ejabberd_sm, Host, ?NS_MAM_06),
-    [mod_disco:unregister_feature(Host, Feature) || Feature <- features(?MODULE, Host)],
+    ?LOG_ERROR(#{what => mam_stopping, host_type => HostType}),
```
Comment (NelsonVides) [inline src/mam/mod_mam.erl:211]: More leftovers. The host_type could actually be a field of the previous LOG_INFO. But surely this shouldn't be an error.

**Review by NelsonVides — COMMENTED** [review]
A couple of left-overs.

File: src/muc_light/mod_muc_light_db_rdbms.erl [hunk 21]
```diff
@@ -71,13 +71,13 @@
 
 %% ------------------------ Backend start/stop ------------------------
 
--spec start(Host :: jid:server(), MUCHost :: jid:server()) -> ok.
-start(_Host, _MUCHost) ->
+-spec start(Host :: jid:server()) -> ok.
+start(_Host) ->
```
Comment (NelsonVides) [inline src/muc_light/mod_muc_light_db_rdbms.erl:75]: Aren't these `HostTypes` instead? If only, for good notation.

File: src/muc_light/mod_muc_light_db_rdbms.erl [hunk 22]
```diff
@@ -71,13 +71,13 @@
...
     prepare_queries(),
     ok.
 
--spec stop(Host :: jid:server(), MUCHost :: jid:server()) -> ok.
-stop(_Host, _MUCHost) ->
+-spec stop(Host :: jid:server()) -> ok.
+stop(_Host) ->
```
Comment (NelsonVides) [inline src/muc_light/mod_muc_light_db_rdbms.erl:80]: As above

File: src/muc_light/mod_muc_light.erl [hunk 23]
```diff
@@ -138,66 +164,60 @@ delete_room(RoomUS) ->
...
 
-    ok.
+set_dynamic_opts(HostType, Opts) ->
+    %% Prepare config schema
+    Def = gen_mod:get_opt(config_schema, Opts, default_schema_definition()),
+    ConfigSchema = mod_muc_light_room_config:schema_from_definition(Def),
+    %% XXX That is a bad style
+    gen_mod:set_module_opt(HostType, ?MODULE, computed_config_schema, ConfigSchema),
+    %% Prepare default config
+    DefaultConfig = mod_muc_light_room_config:default_from_schema(ConfigSchema),
```
Comment (NelsonVides) [inline src/muc_light/mod_muc_light.erl:217]: What is this, and why is it a bad style?

Comment (arcusfelis) [reply src/muc_light/mod_muc_light.erl:217]: modules should not set their own options.

**Review by NelsonVides — APPROVED** [review]
It's impossible to be sure that an almost 3k lines big PR is correct, and that is an amazing way to introduce issues, this is impossible to truly review. Let's see what we come up fixing in the future.

**Comment by chrzaszcz** [comment]
> It's impossible to be sure that an almost 3k lines big PR is correct, and that is an amazing way to introduce issues, this is impossible to truly review. Let's see what we come up fixing in the future.

We should have a second reviewer in such cases.
